ouroboros-consensus-cardano-test/src/Test/Consensus/Cardano/Examples.hs

```diff
@@ -71,29 +73,77 @@ eraExamples =
 
 -- | By using this function, we can't forget to update this test when adding a
 -- new era to 'CardanoEras'.
-combineEras ::
+mkCardanoExamples ::
      NP Examples (CardanoEras Crypto)
   -> Examples (CardanoBlock Crypto)
-combineEras = mconcat . hcollapse . hap eraInjections
+mkCardanoExamples perEraExamples = Golden.Examples {
```

Comment:
Nice! Glad this worked out.

- Instead of the `::` annotations, do you think `@` type applications would be legible enough?

- The `coerce (foo (coerce . blah) ...)` seems off to me. I'm wondering if `coerce foo blah ...` would work instead. But I'm not confident one way or the other. Would you check eg `exampleApplyTxErr       = coerce (viaInject @WrapApplyTxErr) Golden.exampleApplyTxErr`?

Reply:
I originally had to shuffle the types around a lot to get things to work, but indeed I've managed to simplify a bit further as per your suggestions. Let me know if you're still not quite satisfied.

Reply:
Ah, OK, I see now that `viaInject` has a rank2 type, which spoils my idea to use a single `coerce`.

Since I got it into `ghci`, here's what I would suggest.

```
mkCardanoExamples perEraExamples = Golden.Examples {
        exampleBlock            = coerce $ viaInject @I                 (coerce Golden.exampleBlock)
      , exampleSerialisedBlock  =          viaInject                            Golden.exampleSerialisedBlock
      , exampleHeader           =          viaInject                            Golden.exampleHeader
      , exampleSerialisedHeader =          viaInject                            Golden.exampleSerialisedHeader
      , exampleHeaderHash       = coerce $ viaInject @WrapHeaderHash    (coerce Golden.exampleHeaderHash)
      , exampleGenTx            =          viaInject                            Golden.exampleGenTx
      , exampleGenTxId          = coerce $ viaInject @WrapGenTxId       (coerce Golden.exampleGenTxId)
      , exampleApplyTxErr       = coerce $ viaInject @WrapApplyTxErr    (coerce Golden.exampleApplyTxErr)
      , exampleQuery            =          viaInject                            Golden.exampleQuery
      , exampleResult           =          viaInject                            Golden.exampleResult
      , exampleAnnTip           =          viaInject                            Golden.exampleAnnTip
      , exampleLedgerState      =          viaInject                            Golden.exampleLedgerState
      , exampleChainDepState    = coerce $ viaInject @WrapChainDepState (coerce Golden.exampleChainDepState)
      , exampleExtLedgerState   =          viaInject                            Golden.exampleExtLedgerState

      , exampleLedgerConfig = exampleLedgerConfigCardano
      }
  where
    viaInject ::
         forall f. Inject f
      => (forall blk. Examples blk -> Labelled (f blk))
      -> Labelled (f (CardanoBlock Crypto))
```

Nothing else I could come up with was better than what you already have. My only grumble about what you have is that it makes the type of `viaInject` less legible.

I tried writing a bespoke `coerce'` that would handle the rank2 polymorphism of `viaInject` but it's confounded by the fact that some of those `coerce` occurrences above are working on multiple layers of application, so one combinator that abstracts over a single `f` doesn't cut it.

---

Feel free to adopt either what you have or my suggestion in this comment and click Resolve.